corva/models/stream.py

```diff
@@ -1,27 +1,75 @@
-from typing import Any, Dict, List, Optional
+from __future__ import annotations
 
-from corva.models.base import BaseContext, BaseEventData
+from typing import Dict, List, Optional
+
+from pydantic import parse_raw_as
+
+from corva.models.base import BaseContext, BaseEventData, ListEvent
 from corva.state.redis_state import RedisState
 
 
 class StreamContext(BaseContext):
     state: RedisState
 
 
+class RecordData(BaseEventData):
+    hole_depth: Optional[float] = None
+    weight_on_bit: Optional[int] = None
+    state: Optional[str] = None
+
+
 class Record(BaseEventData):
     timestamp: Optional[int] = None
     asset_id: int
     company_id: int
     version: int
-    data: Dict[str, Any]
     measured_depth: Optional[float] = None
     collection: str
+    data: RecordData
 
 
-class StreamEventData(BaseEventData):
-    records: List[Record]
-    metadata: Dict[str, Any]
-    asset_id: int
+class AppMetadata(BaseEventData):
     app_connection_id: int
+
+
+class StreamEventMetadata(BaseEventData):
     app_stream_id: int
-    is_completed: bool
+    apps: Dict[str, AppMetadata]
+
+
+class StreamEventData(BaseEventData):
+    app_key: Optional[str] = None
+    records: List[Record]
+    metadata: StreamEventMetadata
+
+    @property
+    def asset_id(self) -> int:
+        return self.records[0].asset_id
+
+    @property
+    def app_connection_id(self) -> int:
+        return self.metadata.apps[self.app_key].app_connection_id
+
+    @property
+    def app_stream_id(self) -> int:
+        return self.metadata.app_stream_id
+
+    @property
+    def is_completed(self) -> bool:
+        if len(self.records):
```

Comment:
```suggestion
        if self.records:
```

Any reason not to do this?